lib/datasource.js

```diff
@@ -2102,39 +2103,122 @@ DataSource.prototype.copyModel = function copyModel(Master) {
-  transaction.models = transaction.modelBuilder.models;
-  transaction.definitions = transaction.modelBuilder.definitions;
+  var modelBuilder = new ModelBuilder();
+  transactionSource.modelBuilder = modelBuilder;
+  transactionSource.models = modelBuilder.models;
+  transactionSource.definitions = modelBuilder.definitions;
 
-  for (var i in dataSource.modelBuilder.models) {
-    dataSource.copyModel.call(transaction, dataSource.modelBuilder.models[i]);
+  var connector = dataSource.connector;
+  cb = cb || utils.createPromiseCallback();
+  if (connector.beginTransaction) { // Database
+    Transaction.begin(transactionSource.connector, options,
+      function(err, transaction) {
+        transactionSource.transaction = transaction;
+        if (err) {
+          cb(err);
+        } else {
+          if (options.timeout) {
+            setTimeout(function() {
+              var context = {
+                transaction: transaction,
+                operation: 'timeout',
+              };
+              transaction.notifyObserversOf('timeout', context, function(err) {
+                if (!err) {
+                  transaction.rollback(function() {
+                    debug('Transaction %s is rolled back due to timeout',
+                      transaction.id);
+                  });
+                }
+              });
+            }, options.timeout);
+          }
+          var done = function(err) {
+            if (err) {
+              transaction.rollback(function(error) {
+                cb(err || error);
+              });
+            } else {
+              transaction.commit(function(error) {
+                cb(error);
+              });
+            }
+            // make sure cb() isn't called twice, e.g. if `execute` returns a
+            // thenable object and also calls the passed `cb` function
+            done = function() {};
+          };
+          try {
+            var result = execute(transactionSource.models, done);
+            if (result && typeof result.then === 'function') {
+              result.then(function() { done(); }, done);
+            } else if (execute.length < 2) {
+              done();
+            }
+          } catch (err) {
+            done(err);
+          }
+        }
+      }
+    );
+  } else if (connector.transaction) { // Memory, Transient
+    transactionSource.connector = connector.transaction();
+    transactionSource.exec = function(cb) {
+      transactionSource.connector.exec(cb);
+    };
+  } else {
+    throw new Error(g.f('Transactions are not supported'));
   }
 
-  transaction.exec = function(cb) {
-    transaction.connector.exec(cb);
-  };
+  // we can only copy the models after transaction.connector is set up
+  var models = dataSource.modelBuilder.models;
+  for (var i in models) {
+    dataSource.copyModel.call(transactionSource,  models[i]);
+  }
 
-  return transaction;
+  return connector.transaction ? transactionSource : cb.promise;
```

Comment:
Uff, a return type that allows both a transaction and a promise is pretty confusing to me. Are we expecting all users of `ds.transaction` to check what has been returned? I agree backwards compatibility is important, can we perhaps use the number of arguments to decide whether this function was called in the old way (and the transaction should be returned), or in the new way (and the promise should be returned)?